vmware: Packer fails to create the virtual disk if it has spaces in the path

Hi.

If the virtual machine disk path or file name have a space, packer fails to create the virtual disk.

type: vmware-iso
remote_type: esx5
https://github.com/hashicorp/packer/blob/master/builder/vmware/iso/driver_esx5.go

```
: Creating required virtual machine disks
: Error creating disk: 'vmkfstools -c 40960M -d zeroedthick -a lsilogic /vmfs/volumes/MyVolume/Path with spaces/file name with spaces.vmdk'
: Stderr: Extra arguments at the end of the command line.
: Step "stepCreateDisk" failed, aborting...
```
  
Packer v1.2.5
Host: Windows 10 64-bit

Thank you.

Thank you, @SwampDragons.

I just tried this out.

I'm still getting an error message, but it's a little different.

```
2018/10/10 13:54:27 ui: ==> packer test: Creating required virtual machine disks
2018/10/10 13:54:27 packer.exe: 2018/10/10 13:54:27 [INFO] Creating disk with Path: packer test\disk.vmdk and Size: 1024M
2018/10/10 13:54:27 packer.exe: 2018/10/10 13:54:27 [DEBUG] Opening new ssh session
2018/10/10 13:54:27 packer.exe: 2018/10/10 13:54:27 [DEBUG] starting remote command: vmkfstools -c 1024M -d zeroedthick -a lsilogic /vmfs/volumes/MyVolume/packer\ test/disk.vmdk
2018/10/10 13:54:27 packer.exe: 2018/10/10 13:54:27 [ERROR] Remote command exited with '255': vmkfstools -c 1024M -d zeroedthick -a lsilogic /vmfs/volumes/MyVolume/packer\ test/disk.vmdk
2018/10/10 13:54:27 ui error: ==> packer test: Error creating disk: 'vmkfstools -c 1024M -d zeroedthick -a lsilogic /vmfs/volumes/MyVolume/packer\ test/disk.vmdk'
==> packer test: Stderr: Failed to create virtual disk: The system cannot find the file specified (25).
2018/10/10 13:54:27 ui error: ==> packer test: Error creating disk: 'vmkfstools -c 1024M -d zeroedthick -a lsilogic /vmfs/volumes/MyVolume/packer\ test/disk.vmdk'
==> packer test: Stderr: Failed to create virtual disk: The system cannot find the file specified (25).
2018/10/10 13:54:27 ui error: ==> packer test: Step "stepCreateDisk" failed, aborting...
```

It created the directory `/vmfs/volumes/MyVolume/packer` but I expected it to create `/vmfs/volumes/MyVolume/packer test`.

If the `/vmfs/volumes/MyVolume/packer test` directory already exists, then it gets further.

Right, packer.zip from `2018-10-10T22:23:22Z` created the directory `/vmfs/volumes/MyVolume/packer`.

The last packer.zip you sent me also created the directory `/vmfs/volumes/MyVolume/packer` and failed with the same error.

I manually ran `vmkfstools` and it turns out escaping the space with a backslash, using double quotes, or using single quotes will create the virtual disk if the directory already exists.

You are right and packer tries to create the directory before creating the virtual disk.

I would expect that if packer fails to create the directory that it would exit at that point and not continue to try and create the virtual disk.

Here's the full log with the last packer.zip.
[packer.log](https://github.com/hashicorp/packer/files/2470725/packer.log)

Here's a snippet:
```
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [DEBUG] starting remote command: test -e /vmfs/volumes/MyVolume/packer test
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [ERROR] Remote command exited with '2': test -e /vmfs/volumes/MyVolume/packer test
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [DEBUG] starting remote command: mkdir -p /vmfs/volumes/MyVolume/packer test
2018/10/11 14:01:26 ui: ==> packer test: Creating required virtual machine disks
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [INFO] Creating disk with Path: packer test\disk.vmdk and Size: 1024M
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [DEBUG] Opening new ssh session
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [DEBUG] starting remote command: vmkfstools -c 1024M -d zeroedthick -a lsilogic '/vmfs/volumes/MyVolume/packer test/disk.vmdk'
2018/10/11 14:01:26 packer.exe: 2018/10/11 14:01:26 [ERROR] Remote command exited with '255': vmkfstools -c 1024M -d zeroedthick -a lsilogic '/vmfs/volumes/MyVolume/packer test/disk.vmdk'
2018/10/11 14:01:26 ui error: ==> packer test: Error creating disk: 'vmkfstools -c 1024M -d zeroedthick -a lsilogic '/vmfs/volumes/MyVolume/packer test/disk.vmdk''
==> packer test: Stderr: Failed to create virtual disk: The system cannot find the file specified (25).
2018/10/11 14:01:26 ui error: ==> packer test: Error creating disk: 'vmkfstools -c 1024M -d zeroedthick -a lsilogic '/vmfs/volumes/MyVolume/packer test/disk.vmdk''
==> packer test: Stderr: Failed to create virtual disk: The system cannot find the file specified (25).
2018/10/11 14:01:26 ui error: ==> packer test: Step "stepCreateDisk" failed, aborting...
```

Thank you!

--- a/builder/vmware/iso/driver_esx5.go
+++ b/builder/vmware/iso/driver_esx5.go
@@ -47,11 +47,11 @@ func (d *ESX5Driver) Clone(dst, src string, linked bool) error {
 
 func (d *ESX5Driver) CompactDisk(diskPathLocal string) error {
 	diskPath := d.datastorePath(diskPathLocal)
-	return d.sh("vmkfstools", "--punchzero", diskPath)
+	return d.sh("vmkfstools", "--punchzero", fmt.Sprintf("\"%s\"", diskPath))
 }
 
 func (d *ESX5Driver) CreateDisk(diskPathLocal string, size string, adapter_type string, typeId string) error {
-	diskPath := fmt.Sprintf("'%s'", d.datastorePath(diskPathLocal))
+	diskPath := fmt.Sprintf("\"%s\"", d.datastorePath(diskPathLocal))
 	return d.sh("vmkfstools", "-c", size, "-d", typeId, "-a", adapter_type, diskPath)
 }
 
@@ -113,7 +113,7 @@ func (d *ESX5Driver) Destroy() error {
 }
 
 func (d *ESX5Driver) IsDestroyed() (bool, error) {
-	err := d.sh("test", "!", "-e", d.outputDir)
+	err := d.sh("test", "!", "-e", fmt.Sprintf("\"%s\"", d.outputDir))
 	if err != nil {
 		return false, err
 	}
@@ -142,7 +142,7 @@ func (d *ESX5Driver) UploadISO(localPath string, checksum string, checksumType s
 func (d *ESX5Driver) RemoveCache(localPath string) error {
 	finalPath := d.cachePath(localPath)
 	log.Printf("Removing remote cache path %s (local %s)", finalPath, localPath)
-	return d.sh("rm", "-f", finalPath)
+	return d.sh("rm", "-f", fmt.Sprintf("\"%s\"", finalPath))
 }
 
 func (d *ESX5Driver) ToolsIsoPath(string) string {
@@ -450,7 +450,7 @@ func (d *ESX5Driver) CommHost(state multistep.StateBag) (string, error) {
 //-------------------------------------------------------------------
 
 func (d *ESX5Driver) DirExists() (bool, error) {
-	err := d.sh("test", "-e", d.outputDir)
+	err := d.sh("test", "-e", fmt.Sprintf("\"%s\"", d.outputDir))
 	return err == nil, nil
 }
 
@@ -482,11 +482,11 @@ func (d *ESX5Driver) MkdirAll() error {
 }
 
 func (d *ESX5Driver) Remove(path string) error {
-	return d.sh("rm", path)
+	return d.sh("rm", fmt.Sprintf("\"%s\"", path))
 }
 
 func (d *ESX5Driver) RemoveAll() error {
-	return d.sh("rm", "-rf", d.outputDir)
+	return d.sh("rm", "-rf", fmt.Sprintf("\"%s\"", d.outputDir))
 }
 
 func (d *ESX5Driver) SetOutputDir(path string) {
@@ -579,7 +579,7 @@ func (d *ESX5Driver) checkGuestIPHackEnabled() error {
 }
 
 func (d *ESX5Driver) mkdir(path string) error {
-	return d.sh("mkdir", "-p", path)
+	return d.sh("mkdir", "-p", fmt.Sprintf("\"%s\"", path))
 }
 
 func (d *ESX5Driver) upload(dst, src string) error {
@@ -593,7 +593,7 @@ func (d *ESX5Driver) upload(dst, src string) error {
 
 func (d *ESX5Driver) verifyChecksum(ctype string, hash string, file string) bool {
 	if ctype == "none" {
-		if err := d.sh("stat", file); err != nil {
+		if err := d.sh("stat", fmt.Sprintf("\"%s\"", file)); err != nil {
 			return false
 		}
 	} else {

--- a/builder/vmware/iso/driver_esx5.go
+++ b/builder/vmware/iso/driver_esx5.go
@@ -92,7 +92,7 @@ func (d *ESX5Driver) Register(vmxPathLocal string) error {
 	if err := d.upload(vmxPath, vmxPathLocal); err != nil {
 		return err
 	}
-	r, err := d.run(nil, "vim-cmd", "solo/registervm", vmxPath)
+	r, err := d.run(nil, "vim-cmd", "solo/registervm", fmt.Sprintf("\"%s\"", vmxPath))
 	if err != nil {
 		return err
 	}

--- a/builder/vmware/iso/driver_esx5.go
+++ b/builder/vmware/iso/driver_esx5.go
@@ -51,8 +51,7 @@ func (d *ESX5Driver) CompactDisk(diskPathLocal string) error {
 }
 
 func (d *ESX5Driver) CreateDisk(diskPathLocal string, size string, adapter_type string, typeId string) error {
-	diskPath := d.datastorePath(diskPathLocal)
-	diskPath = strings.Replace(diskPath, " ", `\ `, -1)
+	diskPath := fmt.Sprintf("'%s'", d.datastorePath(diskPathLocal))
 	return d.sh("vmkfstools", "-c", size, "-d", typeId, "-a", adapter_type, diskPath)
 }
 

--- a/builder/vmware/iso/driver_esx5.go
+++ b/builder/vmware/iso/driver_esx5.go
@@ -11,6 +11,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"strconv"
 	"strings"
 	"time"
 
@@ -47,11 +48,11 @@ func (d *ESX5Driver) Clone(dst, src string, linked bool) error {
 
 func (d *ESX5Driver) CompactDisk(diskPathLocal string) error {
 	diskPath := d.datastorePath(diskPathLocal)
-	return d.sh("vmkfstools", "--punchzero", fmt.Sprintf("\"%s\"", diskPath))
+	return d.sh("vmkfstools", "--punchzero", strconv.Quote(diskPath))
 }
 
 func (d *ESX5Driver) CreateDisk(diskPathLocal string, size string, adapter_type string, typeId string) error {
-	diskPath := fmt.Sprintf("\"%s\"", d.datastorePath(diskPathLocal))
+	diskPath := strconv.Quote(d.datastorePath(diskPathLocal))
 	return d.sh("vmkfstools", "-c", size, "-d", typeId, "-a", adapter_type, diskPath)
 }
 
@@ -92,7 +93,7 @@ func (d *ESX5Driver) Register(vmxPathLocal string) error {
 	if err := d.upload(vmxPath, vmxPathLocal); err != nil {
 		return err
 	}
-	r, err := d.run(nil, "vim-cmd", "solo/registervm", fmt.Sprintf("\"%s\"", vmxPath))
+	r, err := d.run(nil, "vim-cmd", "solo/registervm", strconv.Quote(vmxPath))
 	if err != nil {
 		return err
 	}
@@ -113,7 +114,7 @@ func (d *ESX5Driver) Destroy() error {
 }
 
 func (d *ESX5Driver) IsDestroyed() (bool, error) {
-	err := d.sh("test", "!", "-e", fmt.Sprintf("\"%s\"", d.outputDir))
+	err := d.sh("test", "!", "-e", strconv.Quote(d.outputDir))
 	if err != nil {
 		return false, err
 	}
@@ -142,7 +143,7 @@ func (d *ESX5Driver) UploadISO(localPath string, checksum string, checksumType s
 func (d *ESX5Driver) RemoveCache(localPath string) error {
 	finalPath := d.cachePath(localPath)
 	log.Printf("Removing remote cache path %s (local %s)", finalPath, localPath)
-	return d.sh("rm", "-f", fmt.Sprintf("\"%s\"", finalPath))
+	return d.sh("rm", "-f", strconv.Quote(finalPath))
 }
 
 func (d *ESX5Driver) ToolsIsoPath(string) string {
@@ -450,7 +451,7 @@ func (d *ESX5Driver) CommHost(state multistep.StateBag) (string, error) {
 //-------------------------------------------------------------------
 
 func (d *ESX5Driver) DirExists() (bool, error) {
-	err := d.sh("test", "-e", fmt.Sprintf("\"%s\"", d.outputDir))
+	err := d.sh("test", "-e", strconv.Quote(d.outputDir))
 	return err == nil, nil
 }
 
@@ -482,11 +483,11 @@ func (d *ESX5Driver) MkdirAll() error {
 }
 
 func (d *ESX5Driver) Remove(path string) error {
-	return d.sh("rm", fmt.Sprintf("\"%s\"", path))
+	return d.sh("rm", strconv.Quote(path))
 }
 
 func (d *ESX5Driver) RemoveAll() error {
-	return d.sh("rm", "-rf", fmt.Sprintf("\"%s\"", d.outputDir))
+	return d.sh("rm", "-rf", strconv.Quote(d.outputDir))
 }
 
 func (d *ESX5Driver) SetOutputDir(path string) {
@@ -579,7 +580,7 @@ func (d *ESX5Driver) checkGuestIPHackEnabled() error {
 }
 
 func (d *ESX5Driver) mkdir(path string) error {
-	return d.sh("mkdir", "-p", fmt.Sprintf("\"%s\"", path))
+	return d.sh("mkdir", "-p", strconv.Quote(path))
 }
 
 func (d *ESX5Driver) upload(dst, src string) error {
@@ -593,7 +594,7 @@ func (d *ESX5Driver) upload(dst, src string) error {
 
 func (d *ESX5Driver) verifyChecksum(ctype string, hash string, file string) bool {
 	if ctype == "none" {
-		if err := d.sh("stat", fmt.Sprintf("\"%s\"", file)); err != nil {
+		if err := d.sh("stat", strconv.Quote(file)); err != nil {
 			return false
 		}
 	} else {

--- a/builder/vmware/iso/driver_esx5.go
+++ b/builder/vmware/iso/driver_esx5.go
@@ -89,18 +89,13 @@ func (d *ESX5Driver) Stop(vmxPathLocal string) error {
 
 func (d *ESX5Driver) Register(vmxPathLocal string) error {
 	vmxPath := filepath.ToSlash(filepath.Join(d.outputDir, filepath.Base(vmxPathLocal)))
-	log.Printf("DEBUGGING 6794 (1): upload path is: %s", vmxPath)
 	if err := d.upload(vmxPath, vmxPathLocal); err != nil {
-		log.Printf("DEBUGGING 6794 (2): upload errored: %s", err.Error())
 		return err
 	}
-	log.Printf("DEBUGGING 6794 (3): Upload succeeded")
 	r, err := d.run(nil, "vim-cmd", "solo/registervm", fmt.Sprintf("\"%s\"", vmxPath))
 	if err != nil {
-		log.Printf("DEBUGGING 6794 (4): VM registration errorred: %s", err.Error())
 		return err
 	}
-	log.Printf("DEBUGGING 6794 (5): Registration succeeded")
 	d.vmId = strings.TrimRight(r, "\n")
 	return nil
 }
@@ -590,7 +585,6 @@ func (d *ESX5Driver) mkdir(path string) error {
 func (d *ESX5Driver) upload(dst, src string) error {
 	f, err := os.Open(src)
 	if err != nil {
-		log.Printf("DEBUGGING 6794 (6): unable to open src file for scp upload")
 		return err
 	}
 	defer f.Close()

--- a/builder/vmware/iso/driver_esx5.go
+++ b/builder/vmware/iso/driver_esx5.go
@@ -89,13 +89,18 @@ func (d *ESX5Driver) Stop(vmxPathLocal string) error {
 
 func (d *ESX5Driver) Register(vmxPathLocal string) error {
 	vmxPath := filepath.ToSlash(filepath.Join(d.outputDir, filepath.Base(vmxPathLocal)))
+	log.Printf("DEBUGGING 6794 (1): upload path is: %s", vmxPath)
 	if err := d.upload(vmxPath, vmxPathLocal); err != nil {
+		log.Printf("DEBUGGING 6794 (2): upload errored: %s", err.Error())
 		return err
 	}
+	log.Printf("DEBUGGING 6794 (3): Upload succeeded")
 	r, err := d.run(nil, "vim-cmd", "solo/registervm", fmt.Sprintf("\"%s\"", vmxPath))
 	if err != nil {
+		log.Printf("DEBUGGING 6794 (4): VM registration errorred: %s", err.Error())
 		return err
 	}
+	log.Printf("DEBUGGING 6794 (5): Registration succeeded")
 	d.vmId = strings.TrimRight(r, "\n")
 	return nil
 }
@@ -585,6 +590,7 @@ func (d *ESX5Driver) mkdir(path string) error {
 func (d *ESX5Driver) upload(dst, src string) error {
 	f, err := os.Open(src)
 	if err != nil {
+		log.Printf("DEBUGGING 6794 (6): unable to open src file for scp upload")
 		return err
 	}
 	defer f.Close()

--- a/communicator/ssh/communicator.go
+++ b/communicator/ssh/communicator.go
@@ -563,6 +563,8 @@ func (c *comm) scpUploadSession(path string, input io.Reader, fi *os.FileInfo) e
 	// The target directory and file for talking the SCP protocol
 	target_dir := filepath.Dir(path)
 	target_file := filepath.Base(path)
+	log.Printf("DEBUGGING 6794 (7): target dir is %s", target_dir)
+	log.Printf("DEBUGGING 6794 (8): target file is %s", target_file)
 
 	// On windows, filepath.Dir uses backslash separators (ie. "\tmp").
 	// This does not work when the target host is unix.  Switch to forward slash

--- a/communicator/ssh/communicator.go
+++ b/communicator/ssh/communicator.go
@@ -563,8 +563,8 @@ func (c *comm) scpUploadSession(path string, input io.Reader, fi *os.FileInfo) e
 	// The target directory and file for talking the SCP protocol
 	target_dir := filepath.Dir(path)
 	target_file := filepath.Base(path)
-	log.Printf("DEBUGGING 6794 (7): target dir is %s", target_dir)
-	log.Printf("DEBUGGING 6794 (8): target file is %s", target_file)
+	// Escape spaces in remote directory
+	target_dir = strings.Replace(target_dir, " ", "\\ ", -1)
 
 	// On windows, filepath.Dir uses backslash separators (ie. "\tmp").
 	// This does not work when the target host is unix.  Switch to forward slash

--- a/communicator/ssh/communicator.go
+++ b/communicator/ssh/communicator.go
@@ -563,13 +563,14 @@ func (c *comm) scpUploadSession(path string, input io.Reader, fi *os.FileInfo) e
 	// The target directory and file for talking the SCP protocol
 	target_dir := filepath.Dir(path)
 	target_file := filepath.Base(path)
-	// Escape spaces in remote directory
-	target_dir = strings.Replace(target_dir, " ", "\\ ", -1)
 
 	// On windows, filepath.Dir uses backslash separators (ie. "\tmp").
 	// This does not work when the target host is unix.  Switch to forward slash
 	// which works for unix and windows
 	target_dir = filepath.ToSlash(target_dir)
+
+	// Escape spaces in remote directory
+	target_dir = strings.Replace(target_dir, " ", "\\ ", -1)
 
 	scpFunc := func(w io.Writer, stdoutR *bufio.Reader) error {
 		return scpUploadFile(target_file, input, w, stdoutR, fi)

